activerecord/lib/active_record/connection_adapters/connection_specification.rb

```diff
@@ -179,7 +179,14 @@ def spec(config)
           end
 
           adapter_method = "#{spec[:adapter]}_connection"
-          ConnectionSpecification.new(spec, adapter_method)
+
+          id ||=
+            if config.is_a?(Symbol)
+              config.to_s
+            else
+              "primary"
```

Comment:
TBH, I am not super happy about this either. I had to do this, so if you do `handler.establish_connection(:readonly)` it would use that key as the spec_id. but also fallback to "primary" if you pass a config hash instead.